libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -1867,4 +1926,28 @@ void PbftManager::syncBlockQueuePush(SyncBlock &&block, dev::p2p::NodeID const &
 
 size_t PbftManager::syncBlockQueueSize() const { return sync_queue_.size(); }
 
+std::unordered_map<trx_hash_t, u256> getAllTrxEstimations(const SyncBlock &sync_block) {
+  std::unordered_map<trx_hash_t, u256> result;
+  for (const auto &dag_block : sync_block.dag_blocks) {
+    const auto &transactions = dag_block.getTrxs();
+    const auto &estimations = dag_block.getEstimations();
+    for (uint32_t i = 0; i < transactions.size(); ++i) {
+      result.emplace(transactions[i], estimations[i]);
+    }
+  }
+  return result;
+}
+
+bool PbftManager::checkBlockWeight(const SyncBlock &blk) const {
+  const auto &trx_estimations = getAllTrxEstimations(blk);
+  u256 total_weight = 0;
+  for (const auto &tx : blk.transactions) {
+    total_weight += trx_estimations.at(tx.getHash());
+  }
+  if (total_weight > config_.gas_limit) {
+    return false;
+  }
+  return true;
+}
+
```

Comment:
why do not we keep estimation as part of trx object? I do not see any use-case keeping it part of Dag

Reply:
By including it in dag block we are picking proposal period and state for which it would be estimated. Because from one block to other estimation could be different